src/rpc_client/src/compactor_client.rs

```diff
@@ -82,56 +89,147 @@ impl GrpcCompactorProxyClientCore {
 #[derive(Debug, Clone)]
 pub struct GrpcCompactorProxyClient {
     pub core: Arc<RwLock<GrpcCompactorProxyClientCore>>,
+    endpoint: String,
 }
 
 impl GrpcCompactorProxyClient {
-    pub fn new(channel: Channel) -> Self {
+    pub fn new(channel: Channel, endpoint: String) -> Self {
         let core = Arc::new(RwLock::new(GrpcCompactorProxyClientCore::new(channel)));
-        Self { core }
+        Self { core, endpoint }
     }
 
-    fn _recreate_core(&self, _channel: Channel) {
-        todo!()
+    async fn recreate_core(&self) {
+        tracing::info!("GrpcCompactorProxyClient rpc transfer failed, try to reconnect");
+        let channel = self.connect_to_endpoint().await;
+        let mut core = self.core.write().await;
+        *core = GrpcCompactorProxyClientCore::new(channel);
+    }
+
+    async fn connect_to_endpoint(&self) -> Channel {
+        let endpoint_str: &'static str = Box::leak(self.endpoint.clone().into_boxed_str());
+        let endpoint = Endpoint::from_static(endpoint_str);
```

Comment:
Shall we use `[from_shared](https://docs.rs/tonic/latest/tonic/transport/channel/struct.Endpoint.html#method.from_shared)` so we don't have to manually leak the endpoint?

Reply:
Make sense.